source/Calamari.Shared/Util/InputSubstitution.cs

```diff
@@ -9,22 +11,19 @@ public static class InputSubstitution
     {
         public static string SubstituteAndEscapeAllVariablesInJson(string jsonInputs, IVariables variables)
         {
-            var tempVariableDictionaryToUseForExpandedVariables = new VariableDictionary();
             var template = TemplateParser.ParseTemplate(jsonInputs);
+            jsonInputs = template.ToString(); // we parse the template back to string to have a consistent representation of Octostache expressions
             foreach (var templateToken in template.Tokens)
             {
-                // TODO: we need change this to have a better way of escaping json string here
                 var arguments = templateToken.GetArguments();
                 if (!arguments.Any()) continue; // to avoid TextToken, which doesn't need to be evaluated
-                var variableName = templateToken.ToString()
-                                                .Replace("#{", string.Empty)
-                                                .Replace("}", string.Empty);
-                var expanded = variables.Evaluate($"#{{{variableName} | JsonEscape}}");
-                tempVariableDictionaryToUseForExpandedVariables.Add(variableName, expanded);
+                string evaluated = variables.Evaluate(templateToken.ToString());
+                if (evaluated == templateToken.ToString())
+                    throw new CommandException($"Variable {templateToken.ToString()} could not be evaluated, check that the referenced variables exist.");
+                jsonInputs = jsonInputs.Replace($"\"{templateToken.ToString()}\"", JsonConvert.ToString(evaluated));
```

Comment:
I don't quite get it here. Why do we include double quotes in the replacement? 
For example, inputs = `{ "ExtractedPath": "#{Package[package].ExtractedPath}" }` will be evaluated to  `{ "ExtractedPath": C:\\some-path }` => double quotes are replaced. Is this what would happen?

Reply:
`JsonConvert.ToString()` adds extra quotes around the result, I couldn't find a way to avoid that.